src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -0,0 +1,481 @@
+                {
+                    continue;
+                }
+
+                // Validate the name and target type match.
+                if (fd.Name.Equals(name)
+                    && fieldType == fd.FieldType)
+                {
+                    context.TargetField = fd;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private static MethodIL GenerateAccessor(ref GenerationContext context)
+        {
+            ILEmitter emit = new ILEmitter();
+            ILCodeStream codeStream = emit.NewCodeStream();
+
+            // Load stub arguments.
+            // When the target is static, the first argument is only
+            // used to look up the target member to access and ignored
+            // during dispatch.
+            int beginIndex = context.IsTargetStatic ? 1 : 0;
+            int stubArgCount = context.Declaration.Signature.Length;
+            for (int i = beginIndex; i < stubArgCount; ++i)
+            {
+                codeStream.EmitLdArg(i);
+            }
+
+            // Provide access to the target member
+            MethodDesc owningMethod;
+            switch (context.Kind)
+            {
+                case UnsafeAccessorKind.Constructor:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.newobj, emit.NewToken(context.TargetMethod));
+                    owningMethod = context.Declaration;
+                    break;
+                case UnsafeAccessorKind.Method:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.callvirt, emit.NewToken(context.TargetMethod));
+                    owningMethod = context.TargetMethod;
+                    break;
+                case UnsafeAccessorKind.StaticMethod:
+                    Debug.Assert(context.TargetMethod != null);
+                    codeStream.Emit(ILOpcode.call, emit.NewToken(context.TargetMethod));
+                    owningMethod = context.TargetMethod;
+                    break;
+                case UnsafeAccessorKind.Field:
+                    Debug.Assert(context.TargetField != null);
+                    codeStream.Emit(ILOpcode.ldflda, emit.NewToken(context.TargetField));
+                    owningMethod = context.Declaration;
+                    break;
+                case UnsafeAccessorKind.StaticField:
+                    Debug.Assert(context.TargetField != null);
+                    codeStream.Emit(ILOpcode.ldsflda, emit.NewToken(context.TargetField));
+                    owningMethod = context.Declaration;
+                    break;
+                default:
+                    Debug.Fail("Unknown UnsafeAccessorKind");
+                    owningMethod = null;
+                    break;
+            }
+
+            // Return from the generated stub
+            codeStream.Emit(ILOpcode.ret);
+            return emit.Link(owningMethod);
```

Comment:
I would expect owningMethod to be `context.Declaration` in all cases. Is there a problem with that?

Reply:
Nope. I was previously hitting an issue with it, but it must have been some other misuse. It works for me now.